src/util/segmentation/operations/correction.js

```diff
@@ -0,0 +1,524 @@
+  const { operationData } = evt;
+
+  if (operationData.segmentationMixinType !== `freehandSegmentationMixin`) {
+    logger.error(
+      `correction operation requires freehandSegmentationMixin operationData, recieved ${
+        operationData.segmentationMixinType
+      }`
+    );
+
+    return;
+  }
+
+  const { pixelData, segmentIndex } = operationData;
+
+  const nodes = snapPointsToGrid(evt);
+
+  if (simpleScissorOperation(nodes, evt, segmentIndex)) {
+    return;
+  }
+
+  const operations = splitLineIntoSeperateOperations(nodes, segmentIndex);
+
+  // Create binary labelmap with only this segment for calculations of each operation.
+  const workingLabelMap = new Uint8Array(pixelData.length);
+
+  operations.forEach(operation => {
+    performOperation(operation, pixelData, workingLabelMap, segmentIndex, evt);
+  });
+}
+
+/**
+ * SnapPointsToGrid - Snap the freehand points to the labelmap grid and attach a label for each node.
+ *
+ * @param  {Object[]} points An array of points drawn by the user.
+ * @param  {UInt16Array} pixelData The 2D labelmap.
+ * @param  {Object} evt The cornerstone event.
+ * @returns {Object[]}
+ */
+function snapPointsToGrid(evt) {
+  const { operationData } = evt;
+  const { pixelData, points } = operationData;
+
+  const { image } = evt.detail;
+  const cols = image.width;
+  const rows = image.height;
+
+  const nodes = [];
+
+  for (let i = 0; i < points.length; i++) {
+    const point = points[i];
+
+    let x = Math.floor(point.x);
+    let y = Math.floor(point.y);
+
+    // Clamp within the confines of the image.
+    x = clip(x, 0, cols - 1);
+    y = clip(y, 0, rows - 1);
+
+    const lastNode = nodes[nodes.length - 1];
+
+    // Skip double counting of closely drawn freehand points.
+    if (lastNode && x === lastNode.x && y === lastNode.y) {
+      continue;
+    }
+
+    nodes.push({
+      x,
+      y,
+      segment: pixelData[y * cols + x],
+    });
+  }
+
+  return nodes;
+}
+/**
+ * SimpleScissorOperation - Check if the operation is a simple scissors
+ * add/remove, and performs it if so.
+ * @param  {Object[]} nodes - The nodes snapped to the grid.
+ * @param  {Object} points - An array of points drawn by the user.
+ * @param  {UInt16Array} pixelData The 2D labelmap.
+ * @param  {Object} evt The cornerstone event.
+ * @param  {number} segmentIndex
+ * @returns {boolean} Returns true if the operation was simple.
+ */
+function simpleScissorOperation(nodes, evt, segmentIndex) {
+  let allInside = true;
+  let allOutside = true;
+
+  for (let i = 0; i < nodes.length; i++) {
+    const node = nodes[i];
+
+    if (node.segment === segmentIndex) {
+      allOutside = false;
+    } else {
+      allInside = false;
+    }
+
+    if (!allInside && !allOutside) {
+      break;
+    }
+  }
+
+  if (allOutside) {
+    logger.warn('The line never intersects a segment.');
+    fillInside(evt);
+
+    return true;
+  } else if (allInside) {
+    logger.warn('The line is only ever inside the segment.');
```

Comment:
Are these really warning level events? Feels more like `debug` or `info`

Reply:
debug yeah.

Reply:
Huh, `debug` isn't a thing, I'll add it to info for now, but we should discuss if `logger.debug` should be a thing.

Reply:
I think warn calls out to `console.warn`. I see these as more _you're clearly not doing something right_ hints that show in dev. Kind of like the `prop-type` errors you see when developing in React.

Good related issues for our `debugger` tool:
- https://github.com/visionmedia/debug/issues/538#issuecomment-398848554
- https://github.com/visionmedia/debug/issues/117#issuecomment-49111501

So we either create different instances of `debug` that are for each level of logging, and then surface that in some way that's a bit easier to  consume, or we more clearly define when/why we add a `debug` call.

@mikehazell's usage for demonstrating cornerstone events being raised is a good one.